README.md
# Flutter Search Panel

[![pub package](https://img.shields.io/badge/pub-1.0.0-orange.svg)](https://pub.dartlang.org/packages/flutter_search_panel)

A Widget to allow users to search from a list of string and make a selection (Similar to a DropDownButtom).

## Installing

```yaml
dependencies:
  flutter_search_panel: ^1.0.1
```

### Import

```dart
import 'package:flutter_search_panel/flutter_search_panel.dart';
```

## How To Use

```dart
List<SearchItem<int>> data = [
  SearchItem(0, 'This'),
  SearchItem(1, 'is'),
  SearchItem(2, 'a'),
  SearchItem(3, 'test'),
  SearchItem(4, '.'),
];

FlutterSearchPanel<int>(
  padding: EdgeInsets.all(10.0),
  selected: 3,
  title: 'Demo Search Page',
  data: data,
  icon: new Icon(Icons.check_circle, color: Colors.white),
  color: Colors.blue,
  textStyle: new TextStyle(color: Colors.white, fontWeight: FontWeight.bold, fontSize: 20.0, decorationStyle: TextDecorationStyle.dotted),
  onChanged: (int value) {
    print(value);
  },
),
```

## Screenshots

![alt text](screenshots/1.png "Search Button") ![alt text](screenshots/2.png "Search Panel")  


## Bugs & Requests

If you encounter any bugs feel free to open an issue. Raise a ticket on github for suggestions. Pull request are also welcome.

### Flutter

For help getting started with Flutter, view our online
[documentation](https://flutter.io/).

For help on editing plugin code, view the [documentation](https://flutter.io/platform-plugins/#edit-code).

## License

MIT License



CHANGELOG.md
## 0.0.1

* Initial commit

## 0.0.2

* Updating docs
* Maintenance (Search page title fix)


## 1.0.0

* First stable release
* Description Update

## 1.0.1

* Added Dynamic data

example/test/widget_test.dart
// This is a basic Flutter widget test.
// To perform an interaction with a widget in your test, use the WidgetTester utility that Flutter
// provides. For example, you can send tap and scroll gestures. You can also use WidgetTester to
// find child widgets in the widget tree, read text, and verify that the values of widget properties
// are correct.

import 'package:flutter/material.dart';
import 'package:flutter_search_panel/flutter_search_panel.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_search_panel_example/main.dart';

void main() {
  testWidgets('Verify Flutter Search Panel', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(new MyApp());

    // Verify that platform version is retrieved.
    expect(
        find.byWidgetPredicate(
          (Widget widget) =>
              widget is FlutterSearchPanel && widget.title.startsWith('Demo Search Page'),
        ),
        findsOneWidget);
  });
}


lib/search_page.dart
import 'package:flutter/material.dart';
import 'search_item.dart';

class SearchPage extends StatefulWidget {
  final List<SearchItem> data;
  final Icon icon;
  final TextStyle textStyle;
  final Color color;
  final EdgeInsetsGeometry padding;
  final String title;

  SearchPage(
      {this.data,
      this.icon,
      this.textStyle,
      this.padding,
      this.color,
      this.title});

  @override
  State<StatefulWidget> createState() {
    return new _SearchPageState();
  }
}

class _SearchPageState extends State<SearchPage> {
  TextEditingController controller = new TextEditingController();
  List<SearchItem> _searchList = [];

  List<SearchItem> _dataList = [];
  @override
  void initState() {
    super.initState();
    _dataList = widget.data;
  }

  @override
  void dispose() {
    _searchList.clear();
    controller.dispose();
    super.dispose();
  }

  _onChange(SearchItem selection) {
    Navigator.pop(context, selection);
  }

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      appBar: new AppBar(
        backgroundColor: widget.color,
        title: new Text(widget.title ?? 'Search Page'),
        elevation: 0.0,
      ),
      body: new Column(
        children: <Widget>[
          new Container(
            color: widget.color,
            child: new Padding(
              padding: widget.padding,
              child: new Card(
                child: new ListTile(
                  leading: new Icon(Icons.search),
                  title: new TextField(
                    controller: controller,
                    decoration: new InputDecoration(
                        hintText: 'Search', border: InputBorder.none),
                    onChanged: onSearchTextChanged,
                  ),
                  trailing: new IconButton(
                    icon: new Icon(Icons.cancel),
                    onPressed: () {
                      controller.clear();
                      onSearchTextChanged('');
                    },
                  ),
                ),
              ),
            ),
          ),
          new Expanded(
            child: _searchList.length != 0 || controller.text.isNotEmpty
                ? new ListView.builder(
                    itemCount: _searchList.length,
                    itemBuilder: (context, i) {
                      return new Card(
                        color: widget.color,
                        child: FlatButton(
                          onPressed: () {
                            Navigator.pop(context, _searchList[i]);
                          },
                          child: new ListTile(
                            leading: widget.icon,
                            title: new Text(_searchList[i].text,
                                style: widget.textStyle),
                          ),
                        ),
                        margin: const EdgeInsets.all(0.0),
                      );
                    },
                  )
                : new ListView.builder(
                    itemCount: _dataList.length,
                    itemBuilder: (context, index) {
                      return new Card(
                        color: widget.color,
                        child: FlatButton(
                            onPressed: () {
                              _onChange(_dataList[index]);
                            },
                            child: new ListTile(
                              leading: widget.icon,
                              title: new Text(
                                _dataList[index].text,
                                style: widget.textStyle,
                              ),
                            )),
                        margin: const EdgeInsets.all(0.0),
                      );
                    },
                  ),
          ),
        ],
      ),
    );
  }

  onSearchTextChanged(String text) async {
    _searchList.clear();
    if (text.isEmpty) {
      setState(() {});
      return;
    }

    _dataList.forEach((userDetail) {
      if (userDetail.text.startsWith(text.toLowerCase()) ||
          userDetail.text.toLowerCase().startsWith(
              text.toLowerCase())) if (_searchList.contains(userDetail) ==
          false) {
        _searchList.add(userDetail);
      }
    });

    setState(() {});
  }
}


lib/flutter_search_panel.dart
import 'package:flutter/material.dart';
import 'search_page.dart';
import 'search_item.dart';

class FlutterSearchPanel<TValue> extends StatefulWidget {
  @required
  final Function onChanged;
  @required
  final List<SearchItem<TValue>> data;
  final String title;
  final Icon icon;
  final TextStyle textStyle;
  final Color color;
  final EdgeInsetsGeometry padding;
  final TValue selected;

  FlutterSearchPanel(
      {this.onChanged,
      this.title,
      this.data,
      this.icon,
      this.textStyle,
      this.selected,
      this.padding,
      this.color});

  @override
  State<StatefulWidget> createState() {
    return new _FlutterSearchPanelState<TValue>();
  }
}

class _FlutterSearchPanelState<TValue>
    extends State<FlutterSearchPanel<TValue>> {
  SearchItem<TValue> selection;
  final _defaultIcon = Icons.label;
  final TextStyle _defaultTextStyle =
      new TextStyle(color: Colors.black, fontSize: 12.0);
  final Color _defaultColor = Colors.white;
  final EdgeInsetsGeometry _defaultPadding = EdgeInsets.all(10.0);

  @override
  void initState() {
    super.initState();
    if (widget.selected != null) {
      selection = widget.data.firstWhere(
          (item) => item.value == widget.selected,
          orElse: () => selection = widget.data[0]);
    } else {
      selection = widget.data[0];
    }
  }

  @override
  void dispose() {
    super.dispose();
  }

  _openSearchPage() async {
    final result = await Navigator.push(
      context,
      MaterialPageRoute(
          builder: (context) => SearchPage(
              title: widget.title,
              data: widget.data,
              icon: widget.icon ?? _defaultIcon,
              padding: widget.padding ?? _defaultPadding,
              color: widget.color ?? _defaultColor,
              textStyle: widget.textStyle ?? _defaultTextStyle)),
    );

    if (result != null) {
      setState(() {
        selection = result;
      });
      if (widget.onChanged != null) {
        widget.onChanged((result as SearchItem<TValue>).value);
      }
    }
  }

  @override
  Widget build(BuildContext context) => new RaisedButton(
        color: widget.color ?? Colors.blue,
        child: Flex(
          direction: Axis.horizontal,
          mainAxisSize: MainAxisSize.min,
          children: <Widget>[
            Flexible(
              child: Padding(
                padding: const EdgeInsets.all(10.0),
                child: Text(
                  selection.text,
                  style: widget.textStyle ?? Theme.of(context).textTheme.button,
                ),
              ),
            ),
          ],
        ),
        padding: widget.padding,
        onPressed: _openSearchPage,
      );
}


android/src/main/java/xyz/altercode/fluttersearchpanel/FlutterSearchPanelPlugin.java
package xyz.altercode.fluttersearchpanel;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** FlutterSearchPanelPlugin */
public class FlutterSearchPanelPlugin implements MethodCallHandler {
  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "flutter_search_panel");
    channel.setMethodCallHandler(new FlutterSearchPanelPlugin());
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    } else {
      result.notImplemented();
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_search_panel/flutter_search_panel.dart';
import 'package:flutter_search_panel/search_item.dart';

void main() => runApp(new MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => new _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    List<SearchItem<int>> data = [
      SearchItem(0, 'This'),
      SearchItem(1, 'is'),
      SearchItem(2, 'a'),
      SearchItem(3, 'test'),
      SearchItem(4, '.'),
    ];

    return new MaterialApp(
      home: new Scaffold(
        appBar: new AppBar(
          title: const Text('Flutter Search Panel'),
        ),
        body: Center(
          child: FlutterSearchPanel<int>(
            padding: EdgeInsets.all(10.0),
            selected: 2,
            title: 'Demo with Integer',
            data: data,
            icon: new Icon(Icons.check_circle, color: Colors.white),
            color: Colors.blue,
            textStyle: new TextStyle(
                color: Colors.white,
                fontWeight: FontWeight.bold,
                fontSize: 20.0,
                decorationStyle: TextDecorationStyle.dotted),
            onChanged: (int value) {
              print(value);
            },
          ),
        ),
      ),
    );
  }
}
